File: Sources/UITesting/Asserions/Value.swift
import XCTest

extension State {

  @discardableResult
  public consuming func expect<Value: Equatable>(
    _ select: (Content) -> Query<Value>,
    is expected: Value
  ) throws -> Self {
    let query = select(content)
    let value = query(application)
    guard value == expected else {
      throw IncorrectValue(value: value, expected: expected)
    }
    return self
  }
}

// MARK: IncorrectValue

@MainActor
struct IncorrectValue<Value>: Error {
  fileprivate let value: Value
  fileprivate let expected: Value
}

extension IncorrectValue: @MainActor CustomStringConvertible {
  var description: String {
    "Value is incorrect. Was \(value) expected: \(expected)"
  }
}

File: src/main/kotlin/com/marksandspencer/uitesting/assertions/Value.kt
package com.marksandspencer.uitesting.assertions

import com.marksandspencer.uitesting.Query
import com.marksandspencer.uitesting.Screen
import com.marksandspencer.uitesting.State

fun <Content : Screen, Value> State<Content>.expect(
    query: Content.() -> Query<Value>,
    isEqualTo: Value,
): State<Content> {
    val q = content.query()
    val value = q(compose)
    if (value != isEqualTo) {
        throw IncorrectValue(value, isEqualTo)
    }
    return this
}

class IncorrectValue internal constructor(
    val actual: Any?,
    val expected: Any?,
) : AssertionError("Value is incorrect. Was $actual expected: $expected")
